packages/signals/src/impls.rs

```diff
@@ -279,16 +280,51 @@ impl<T: Clone + 'static> IntoIterator for Signal<Vec<T>> {
     }
 }
 
+/// An iterator over items in a `Signal<Vec<T>>` that yields [`MappedSignal`]s.
+pub struct MappedSignalIterator<T: 'static> {
+    index: usize,
+    length: usize,
+    value: Signal<Vec<T>>,
+}
+
+impl<T> Iterator for MappedSignalIterator<T> {
+    type Item = MappedSignal<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let index = self.index;
+        self.index += 1;
+        (index < self.length).then(|| self.value.map(move |v| v.get(index).unwrap()))
+    }
+}
+
 impl<T: 'static> Signal<Vec<T>> {
     /// Returns a reference to an element or `None` if out of bounds.
     pub fn get_mut(&self, index: usize) -> Option<Write<T, Vec<T>>> {
         Write::filter_map(self.write(), |v| v.get_mut(index))
     }
+
+    /// Create an iterator of [`MappedSignal`]s over the inner vector.
+    pub fn iter_signals(&self) -> MappedSignalIterator<T> {
+        MappedSignalIterator {
+            index: 0,
+            length: self.read().len(),
+            value: *self,
+        }
+    }
 }
 
 impl<T: 'static> Signal<Option<T>> {
     /// Returns a reference to an element or `None` if out of bounds.
     pub fn as_mut(&self) -> Option<Write<T, Option<T>>> {
         Write::filter_map(self.write(), |v| v.as_mut())
     }
+
+    /// Try to create a [`MappedSignal`] over the inner value.
+    pub fn as_mapped_ref(&self) -> Option<MappedSignal<T>> {
+        if self.read().is_some() {
+            Some(self.map(|v| v.as_ref().unwrap()))
```

Comment:
So yeah, this line can crash. https://github.com/DioxusLabs/dioxus/pull/1537#issuecomment-1873502974

(I figured I'd thread my replies under here so I top spamming the main PR)

Reply:
Yeah... I think (part of) the problem is that changes get *immediately* pushed over to the receiving signal, rather than making effective use of the tree of scopes you guys have got going on.

Perhaps when a signal is written to, it could be marked "dirty", and downstream subscribers pull the value only once they need it. If the scope is dropped before then, that read never happens, the pull never takes place, and the `.unwrap()` (though it could maybe be `.expect()` at least) never panics.

Reply:
Looking more at it, I think you have a bit of a weirdly inconsistent way of implementing push/pull implementation that's leading to the crash I'm facing.

Note how here, when calling `MappedSignal::read()`, you pull the value from the original signal through the mapping into wherever the read occurs.

```
/// Get the current value of the signal. This will subscribe the current scope to the signal.
    pub fn read(&self) -> GenerationalRef<U> {
        (self.mapping.read())()
    }
```

Yet, when an update is available, the information is actively pushed into the subscribing signals, and when they read() it, they just get the value that's just been delivered there.

Reply:
It'd probably require a change to how signals work in general, actually... Have the initial update only propagate a "dirty" status in the dependent signals, and have the signal only be actually updated when `.read()` is called.

Reply:
> It'd probably require a change to how signals work in general, actually... Have the initial update only propagate a "dirty" status in the dependent signals, and have the signal only be actually updated when .read() is called.

For selectors, we need to rerun the selector after any dependency changes because we need to rerun any subscribers that read the selector (like the component) if the value changes

Reply:
I'm not sure if there is a way to prevent that error. I think fundamentally, you can't do a check once and then assume that check still holds the next time you read the mapped value (especially now that signals can be `Sync+Send`. I removed the iterator and Option helpers that map signals based on some initial property.